adam_core/propagator/pyoorb.py

```diff
@@ -444,38 +444,84 @@ def _generate_ephemeris(self, orbits: OrbitType, observers: Observers) -> Epheme
                 tolerance=0.001,  # FIXME: This is 0.001 days, which is 86.4 seconds, not 1 microsecond?
             )
 
-            ephemeris = Ephemeris.from_kwargs(
-                orbit_id=orbit_ids,
-                object_id=object_ids,
-                coordinates=SphericalCoordinates.from_kwargs(
-                    time=Timestamp.from_mjd(
-                        ephemeris[:, 0],
-                        scale="utc",
+            if isinstance(orbits, Orbits):
+
+                # Map the object and orbit IDs back to the input arrays
+                orbit_ids = orbits.orbit_id.to_numpy(zero_copy_only=False)[
+                    orbit_ids_idx
+                ]
+                object_ids = orbits.object_id.to_numpy(zero_copy_only=False)[
+                    orbit_ids_idx
+                ]
+
+                ephemeris = Ephemeris.from_kwargs(
```

Comment:
It seems like we are retyping this variable multiple times? First whatever comes back from pyoorb, then a 2D numpy array, and now our Ephemeris / VariantEphemeris classes. Can we split them out into separate variables? Especially from mutable to immutable.

Reply:
Done. Spit it into `ephemeris_array_3D`, `ephemeris_array`,  and `ephemeris` .